Implement complete end-to-end namespace-based export control

krunaljain opened the pull request:
This PR implements comprehensive namespace-based export control ClusterMesh resources with extending the original clustermesh-apiserver-only implementation to provide complete end-to-end functionality including proper handling of filtering deactivation edge cases and E2E testing.

## Implementation Approach

**ClusterMesh Resource Filtering (clustermesh-apiserver):**
- Added namespace indexing for CiliumIdentity, CiliumEndpoint, and CiliumEndpointSlice
- Namespace watcher processes annotation changes and syncs resources to/from etcd
- When namespace marked global: Resources added to etcd via converters
- When namespace marked local: Resources removed from etcd
- **Edge case handling**: When no annotated namespaces remain, automatically backfills all resources to revert to backwards-compatible behavior

CFP: https://github.com/cilium/design-cfps/pull/74

gandro commented on install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:
```diff
@@ -231,6 +231,12 @@ spec:
+          valueFrom:
+            configMapKeyRef:
+              name: cilium-config
+              key: clustermesh-default-global-namespace
```
This looks incomplete. Where's that key defined, what Helm value do I need to provide to set this?

krunaljain replied:
Will update this

gandro commented on test/k8s/net_policies.go:
```diff
@@ -1280,6 +1280,465 @@ var _ = SkipDescribeIf(func() bool {
+			res.ExpectSuccess("Should be able to remove global annotation from %s", globalNS2)
+
+			// Wait a moment for the changes to propagate
+			time.Sleep(5 * time.Second)
```
Please use a polling-based approach like `Eventually` here (and in all other instance of using `time.Sleep`) and check that the changes actually have been propagated (by inspecting cilium-agent state or similar, or just running the curl in a loop until it gives the expected return code). 

Timing is unreliable in CI  and this will eventually not wait long enough for the change to propagate.

krunaljain replied:
Noted

gandro commented on test/k8s/net_policies.go:
```diff
@@ -1280,6 +1280,465 @@ var _ = SkipDescribeIf(func() bool {
 		})
 	})
 
+	Context("Global Namespace Network Policies", func() {
```
Please be advised that we are trying to remove from Ginkgo and that this test file might be removed in the future. It would be much better if you added a test to cilium-cli instead, in particular since we also run that one in clustermesh configurations.

MrFreezeex replied:
There's also hive script (https://docs.cilium.io/en/latest/contributing/development/hive/) which is already tested here https://github.com/cilium/cilium/tree/main/clustermesh-apiserver/clustermesh/testdata which could be useful to test things here.

For cilium-cli tests might also be nice I am not entirely sure how it would look like though :thinking:, maybe adding some option to set global namespace default to false in one of the clustermesh conformance test and in the CLI connectivity test whenever we are doing a clustermesh test always add `clustermesh.cilium.io/global=true` on the created namespace?

krunaljain replied:
@MrFreezeex will follow the recommended approach

gandro reviewed (commented):
Thanks for your contribution! I did do a first look in my codeowners (the testing and Helm) and left a bit of feedback

MrFreezeex commented on clustermesh-apiserver/clustermesh/cells.go:
```diff
@@ -140,3 +150,19 @@ var pprofConfig = pprof.Config{
+		}
+	}
+	
+	return NamespaceWatcherConfig{DefaultGlobalNamespace: defaultGlobal}
```
I am not entirely sure why we indeed pass cluster id (and other things) as env here https://github.com/cilium/cilium/blob/8d1da44eae344684d599e4f1a77be6db7b302917/install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml#L228 (cc @cilium/sig-clustermesh anyone knows?), but they seems kind of unused. So I would expect this setting to be added in some hive config instead of env variable

MrFreezeex commented on clustermesh-apiserver/clustermesh/namespace_watcher.go:
```diff
@@ -0,0 +1,370 @@
+	} else {
+		nw.annotated.Delete(ns.Name)
+		nw.global.Delete(ns.Name)
+	}
```
Do you have a specific reason to maintain a separate state vs using the namespace store as is to check if namespaces are annotated?

krunaljain replied:
This was mainly leveraged for backfill operations but we can leverage the newly created namespace index for that case. I will remove this

MrFreezeex commented on clustermesh-apiserver/clustermesh/synchronizer.go:
```diff
@@ -116,3 +130,95 @@ func RegisterSynchronizer[T runtime.Object](in syncParams[T]) {
+}
+
+// resourceProcessor implements NamespaceProcessor for a specific resource type
+type resourceProcessor[T runtime.Object] struct {
```
Could you name this `namespaceProcessor` instead? `resourceProcessor` looks too generic and is hard to spot that it's only about filtering namespace IMO

krunaljain replied:
Ack

MrFreezeex commented on clustermesh-apiserver/clustermesh/converters.go:
```diff
@@ -218,12 +257,20 @@ func newCiliumEndpointSliceOptions(cfg cmk8s.CiliumEndpointSliceConfig) Options[
-func ciliumEndpointSliceMapper(endpointslice *cilium_api_v2a1.CiliumEndpointSlice) iter.Seq[store.Key] {
+func ciliumEndpointSliceMapper(endpointslice *cilium_api_v2a1.CiliumEndpointSlice, nsFilter NamespaceFilter) iter.Seq[store.Key] {
 	return func(yield func(store.Key) bool) {
+		// Only export endpoint slices from global namespaces
```
nit so that it doesn't get confused with regular k8s epslice
```suggestion
		// Only cilium export endpoint slices from global namespaces
```

MrFreezeex commented on clustermesh-apiserver/clustermesh/synchronizer.go:
```diff
@@ -116,3 +130,95 @@ func RegisterSynchronizer[T runtime.Object](in syncParams[T]) {
+// Only CiliumIdentity, CiliumEndpoint, and CiliumEndpointSlice need namespace processors.
+func needsNamespaceProcessor(resourceName string) bool {
+	switch resourceName {
+	case "CiliumIdentity", "CiliumEndpoint", "CiliumEndpointSlice":
```
There is ServiceExport sync (mcs) too https://github.com/cilium/cilium/blob/main/pkg/clustermesh/mcsapi/serviceexportsync.go + https://github.com/cilium/cilium/blob/main/operator/watchers/service_sync.go that also should get the filtering.

Commenting here but if those resources doesn't get added specifically in this function (which might makes sense?) the comment should at least says that this isn't the only resources subject to the namespace processor / only affect resources only exported via the clustermesh-apiserver

krunaljain replied:
@MrFreezeex the MCS changes will come in a follow up PR. Want to break down the functionality into multiple PRs to ease up reviews

MrFreezeex replied:
Note the second link in my first message in this thead is not MCS specific it's about global service export (sorry if that wasn't clear).

I think it might help to review this together since we would be able to review how you would handle change outside the clustermesh-apiserver packages (also note that those two export code might run in the operator or the clustermesh-apiserver depending on the config used).  And FYI the code to sync to the clustermesh-apiserver for MCS is pretty much isolated / doesn't interact with the rest of the code in the mcsapi package and is quite similar to what's done here.

krunaljain replied:
@MrFreezeex wanted to clarify this comment before publishing the next diff. I was thinking of breaking down the PRs into three components 
1)CiliumIdentity/CiliumEndpoint/CiliumEndpointSlice exports  
2) Global Service Exports 
3) MCS exports 
Do you prefer having all these components in a single PR?

MrFreezeex replied:
I think it could be ok to split as long as there is  at least one of those PR also open so that we could check that the API makes sense for a usage outside the clustermesh-apiserver go packages

The same PR for everything could be ok too though (excluding the controller change for MCS which you already have a draft PR opened for it).

krunaljain replied:
Ack. I will restrict it to two PRs in that case. One with the MCS changes and the other containing everything else

krunaljain replied:
@joestringer @MrFreezeex we had a discussion on this thread to break the CFP down into two PRs (One dedicated to MCS and the other one containing rest of the logic). Is the preference to break it down into more PRs? I can break the functionality down into 4-5 individual PRs but wanted to get alignment before doing so.

MrFreezeex replied:
My intent is mainly that you can demonstrate that you are able to apply this namespace filtering approach/infra in other package than the clustermesh-apiserver and I don't believe it should be that much more code so doing that here looks more practical to me. This doesn't necessarily include docs change though for instance.

One PR doesn't mean one commit though, it would be best that you still break this down in multiple well structured commit!

krunaljain replied:
@MrFreezeex I can split this PR up into more granular commits. Other components (I think it's only the operator) would reuse the namespace_watcher util which resides in the shared pkg/ directory and maintain their individual namespace indexes so there shouldn't be any dependency on the clustermesh-apiserver. With that state being separate, my thinking was we can split that off as an isolated PR. I can add it to this PR itself though but that'll further increase the length and complexity of the overall PR in general although that will be restricted to a separate commit.

MrFreezeex replied:
I think we are talking about different things. I am mainly talking about the parts (with MCS and global service linked in my first message in this thread) that sync/export to the clustermesh-apiserver/kvstore not the parts that use/pull those info.

I would expect that we are able to reuse most of the code for exporting at least, the rest I haven't yet looked at the code you wrote here but it looks fine to split that off in another PR indeed!

joestringer replied:
Splitting PRs is as much an art as it is a science, so it's tough to make blanket judgements. LFX only tracks categories up to 500 LoC but the [review lead time on large PRs scales much longer based on PR size](https://insights.linuxfoundation.org/project/cilium/development?widget=review-time-by-pull-request-size), from 2-3 days for <60LoC up to 8 days average for 500+. When you create a large PR, it is asking reviewers to spend a lot of time on your PRs, and generally review is the bottleneck to making progress.

I can say that when I see PRs <= 500LoC, I think they are digestible and I'm unlikely to be scared away from reviewing. If I see >= 1KLoC and it's not just because there's a big dependency change, my initial inclination is to push the review out later because I can't fit a quick review into a short period. Also worth noting that if there's 3KLoC and you're receiving 10+ comments from each reviewer each time, the PR gets weighed down in whether the feedback is addressed. It's just a lot of churn to handle all at once.

The other thing I'm thinking about here is that it's best if you submit some small bugfixes or small refactoring to improve the quality of the codebase before tackling a large feature like this. Some of the feedback you'll be receiving on this PR is common community culture and ettiquette for submitting PRs, which you could learn more quickly on smaller PRs. Right now because you haven't submitted any prior work, there's also additional burden on reviewers/maintainers to run the tests for you. Combine this with a large PR that receives a lot of feedback, and there's a lot of additional workload that can cause a submission like this to take a long time and tire everyone out. If there are many smaller things that don't match the way we do things, this can distract reviewers from the big picture of what we're trying to achieve because the quality of the PR doesn't match reviewer expectations.

krunaljain replied:
I think if we use LoC as a metric, the only way to alleviate the complexity is by breaking it down into smaller PRs. Splitting it into logical commits would not reduce the complexity of the overall PR in general. I am fine progressing along that route unless there are any concerns otherwise.

joestringer replied:
SGTM yeah. I know that reviewers can sometimes want to know what the "big picture" is when there are smaller PRs, so that's the tradeoff. That said, I think it's OK to submit smaller PRs and reference the CFP to describe why you're trying to make that smaller change. Another option is to just have the bigger PR with all the changes posted as "draft" and point reviewers to that version if they ask for where the overall effort is going. This is where breaking up commit by commit can make the maintenance of multiple PRs a bit easier.

joestringer replied:
LoC is definitely not a perfect metric, but I think it's somewhat representative of complexity. And it's that complexity of all the changes that slows things down, as it's just harder for reviewers to reason about all the possible implications of various changes.

krunaljain replied:
Sounds good. I will break this down into individual PRs in that case.

MrFreezeex replied:
This SGTM in general too, but just to talk more specifically about this change. I started this thread initially because there are two parts in the code with very similar behavior that was not included/still not included in the current PR and reusing the namespace watching infrastructure that you are introducing now doesn't look (at first glance at least) easily extendable to those while it should probably be at least in some capacity.

So splitting in multiple PR is ok but as @joestringer just said there should be one draft PR where we can see the global picture or that it become obvious that this can be reused easily (through some hive Cell for instance or separating the namespace watching in some go package that can be imported in all those places probably).

krunaljain replied:
@MrFreezeex the namespace_watcher module currently resides in pkg/clustermesh which should be sharable across all components? We can export it into a hive config as well but I thought it might fall into the second category you described?

MrFreezeex replied:
Well currently you have two namespace watchers and IIUC it's the one in `clustermesh-apiserver/clustermesh` package that does the relevant pieces to what we are talking about here (at least last I checked this 3 days ago). FYI `pkg/clustermesh` is primarily used for things inside the cilium-agent/DaemonSet and not a common thing, pretty much nothing import this package ATM.

krunaljain replied:
Understood. We can modularize that part into a hive cell to be reused across components.

joestringer replied:
Incidentally, I would have expected Cilium to already have some namespace watcher code in the agent for determining namespace labels. Not sure whether that code is easy to share or not but it might be something to consider.

krunaljain replied:
Will take a look into it. As per the discussion in this thread, I think we agreed to one draft PR containing E2E changes and a set of modularized PRs splitting it down into smaller logical components. With that, I can update this PR to serve as the draft PR. Let me know if you have any objections with the same.

MrFreezeex reviewed (changes requested):
Thanks for initiating this! I left a few additional comments.

Also note that your commit doesn't get squashed when merged so typically we try to have clean git history in PR and you can/it's appreciated to have multiple ones for large changes with a description if necessary.

MrFreezeex commented on clustermesh-apiserver/clustermesh/namespace_watcher.go:
```diff
@@ -0,0 +1,370 @@
+const (
+	// GlobalNamespace is the annotation used to mark namespaces for global export in ClusterMesh
+	GlobalNamespaceAnnotation = "clustermesh.cilium.io/global"
+)
```
We typically centralize all the annotations in `pkg/annotation/k8s.go`, could you also do that here :pray:?

krunaljain replied:
Ack

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+
+   .. code-block:: shell-session
+
+      cilium clustermesh status --context $CLUSTER1
```
How is this verifying backwards compatibility?

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+
+3. Check that only resources from global namespaces appear in other clusters
+
+4. Validate global service functionality with both namespace and service annotations
```
The average user is not going to be able to follow these steps. Consider expanding them to instruct the user exactly what commands to run, and provide example output.

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+conditions to be considered a Global Service:
+
+1. The service must be annotated with ``service.cilium.io/global: "true"``
+2. The service must reside within a namespace marked as global
```
This seems like an unordered list:

```suggestion
* The service must be annotated with ``service.cilium.io/global: "true"``
* The service must reside within a namespace marked as global
```

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+     --set clustermesh.defaultGlobalNamespace=false
+
+When set to ``false`` (recommended for security), unannotated namespaces are
+treated as local by default when namespace filtering is active.
```
The `(recommended for security)` aside here is terse and implies a lot more knowledge that is not being described. I would suggest removing that and reconsidering where in this document it makes sense to explain the top level security implications instead.

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+By default, all namespaces in a cluster are considered global for ClusterMesh
+resource sharing. To enhance security and control which namespaces export
+resources to other clusters, you can enable namespace-based export control
+through namespace annotations.
```
I would suggest adding another sentence after this that describes concisely what namespace-based export control does. This paragraph currently implies what it does but doesn't outright describe the feature.

The "To enhance security" part also comes across at least to me as marketing the feature, not describing to the user exactly what the implications are.

joestringer replied:
^ Maybe I need to take another look at the CFP, but worst case I could see this making security worse if users are not expecting the behavior. It's a different model for how information is shared which has certain properties, so I'm not sure it's fair to the user to portray the feature as just improving security. Furthermore, the CFP didn't describe anything about improving security, instead it described improving _scalability_ so I'm surprised to see these documents touting security as the goal.

joestringer commented on .github/workflows/conformance-clustermesh.yaml:
(unrelated to this file)

Please add a reference to the PR description to link to the issue and/or CFP so reviewers can have that context. This is to implement https://github.com/cilium/cilium/issues/39876, right?

joestringer commented on NAMESPACE_EXPORT_CONTROL_TESTING.md:
Please don't litter the top level of the tree for random features.

If you must, consider handwriting a file in the developer documentation pages. Otherwise I'd just suggest deleting this file. There's no way we're going to be able to maintain it and keep it up to date.

joestringer commented on .github/workflows/conformance-clustermesh.yaml:
One more general PR context: In Cilium we expect every commit to compile and run, in order to ensure the tree is bisectable. Please avoid adding commits like "address comments" and instead fold the changes into the individual meaningful commits. 

There's plenty of guidance out there but this classic one for writing good commit messages is a reasonable resource:
https://github.com/torvalds/subsurface-for-dirk/blob/master/README.md#contributing

joestringer commented on Documentation/network/clustermesh/clustermesh.rst:
```diff
@@ -288,8 +288,9 @@ Next Steps
- * :ref:`gs_clustermesh_services`
- * :ref:`gs_clustermesh_network_policy`
+ * :ref:`gs_clustermesh_services` - Learn about global services and load-balancing
+ * :ref:`namespace_export_control` - Configure namespace-based export control for enhanced security
```
nit: Similar earlier feedback, IIRC this wasn't about enhancing security but rather about improving scalability.

```suggestion
 * :ref:`namespace_export_control` - Configure namespace-based export control
```

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
 
+.. _namespace_export_control:
+
+Namespace-based Export Control
```
nit, I didn't really pay close attention to this during the CFP but I think the name could be improved. "Export Control" has another meaning which would be ideal to avoid conflict with. Perhaps...?

```suggestion
.. _namespace_export_control:

Clustermesh Namespace Export Filtering
```

MrFreezeex replied:
Super high level comment but I have two main point of concern about this doc:
- This pages is for Global Services while the global namespace annotation is not tied specifically to (Global) Services, so it should probably be somewhere else. I am not sure where that somewhere else would be though, maybe it could be its own page?
- This documentation looks a bit too verbose as is, but I see that Joe left many comments that would most likely already involve this point too

And as Joe also said, this could very well be in a separated PR!

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+  the traditional ClusterMesh setup (backwards compatible)
+* When the first namespace receives the ``clustermesh.cilium.io/global``
+  annotation, namespace filtering becomes active
+* Only resources from namespaces marked as global will be shared across clusters
```
These are not prerequisites. It's better to just outright describe this in regular paragraphs rather than instructing the person to know something. The reader is already here, let's just describe the functionality directly to them instead of telling them they need to know stuff and listing it.

I suspect this might even work if we just drop the line 78 here and remove the list item formatting and just put the sentences together. It could be wordsmithed to read a bit more naturally than that, but it'd be a good start.

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+1. The service must be annotated with ``service.cilium.io/global: "true"``
+2. The service must reside within a namespace marked as global
+
+.. code-block:: yaml
```
What is this code block? There is no text that describes what it's doing here or what the reader should know.

joestringer commented on Documentation/network/clustermesh/services.rst:
```diff
@@ -62,6 +62,133 @@ Below example will expose remote endpoint without sharing local endpoints.
+  its annotation, filtering becomes inactive and all namespace resources
+  are backfilled into etcd (backwards compatibility mode)
+* **Transition**: Resources are dynamically added/removed from etcd as
+  namespace annotations change
```
Is this information different from the "Prerequisites" information? Seems like this would be better documented together as part of the "concepts" preparation to directly describe the feature.

joestringer commented on contrib/testing/kind-clustermesh1-global-ns.yaml:
I don't see these connected in with any existing infrastructure, how would one use this?

If we add these into the tree, I might suggest putting these two `contrib/` changes into a separate commit and then describe in the commit message how another developer can use it. 

Even better if we could avoid this, though. There are hundreds of flags in Cilium, and we can't realistically maintain hundreds of combinations of values files. If this is a super common case then I think it may make sense to put infrastructure in the tree to simplify setup for this, but I wonder if it'd be better for us to just build an abstraction in the makefiles that allows you to add your own custom values file by doing something like `make kind-clustermesh EXTRA_VALUES=my_values.yaml`, then just don't commit the values file.

joestringer commented on examples/kubernetes/connectivity-check/clustermesh.cue:
I appreciate you putting the time into implementing this CUE, but we don't really extend these any more. I'd suggest focusing on `cilium-cli` tests instead. Please remove this.

joestringer commented on install/kubernetes/cilium/values.yaml:
You need to extend `install/kubernetes/cilium/values.yaml.tmpl`. This file is automatically generated.

joestringer commented on clustermesh-apiserver/clustermesh/testdata/namespace-filtering.txtar:
Isn't this the same as the other txtar in this directory?

What I think would be useful is to see tests where k8s resources are created, then the test validates that the resources are / are not filtered out by this feature.

joestringer commented on clustermesh-apiserver/clustermesh/namespace_watcher_test.go.bak:
Leftover? Please delete it.

Please consider doing a `git show --stat` for your commits at the very least to ensure there aren't extra unneeded files. (I'd suggest doing a full scan over your changes in detail as well to ensure you can convince yourself each line is needed as a self-review, as this can sometimes catch things that will otherwise occupy reviewers' time).

joestringer commented on test/k8s/net_policies.go:
Please remove these test changes - [Ginkgo is now deprecated for writing new testing](https://docs.cilium.io/en/stable/contributing/testing/e2e_legacy/#end-to-end-testing-framework-legacy).

Better is hive-based testing which I see you have already started with. If really needed, end-to-end tests can be added to cilium-cli as well, but these should be used sparingly --- we have many combinations of tests already, and end-to-end tests are hard to make stable.

joestringer replied:
Incidentally, @krunaljain I think we have an opportunity here to improve the experience for new developers. Evidently you discovered these tests and thought that it would make sense to extend them, so the link that I shared above in the docs was not super clear to you.

We have some tech debt here to remove this old testing infrastructure (See https://github.com/cilium/cilium/issues/37837) which we're gradually working our way through. But my question to you would be, do you think we could make it clearer that these tests shouldn't be extended? Any ideas how we could make that more clear to newer contributors like yourself? Was there a particular documentation page you were following for instance which could have a clearer warning not to write logic here?

krunaljain replied:
@joestringer Maybe that page can be linked to the setup page? 
https://docs.cilium.io/en/stable/contributing/development/dev_setup/  so first time developers are aware of the dependencies planned to be deprecated?

joestringer reviewed (changes requested):
Thanks for the PR. As mentioned by other reviewers, Please ensure the related issues and CFPs are linked in the PR description for reviewer reference.

I focused mainly on my codeowners and glanced at a few things outside. Deferring the review of core functionality to the relevant clustermesh owners.

This PR is also huge - 3K+ lines. I would advise at least breaking this up into smaller commits on functional boundaries. If there are smaller functional changes, such as writing a namespace watcher into clustermesh-apiserver, I would suggest submitting that first as a separate PR and highlight that the functionality is needed for CFP-XXXX in that PR. If there are other substantial chunks of functionality on the order of 200-500 LoC, consider doing this for those as well. The documentation changes would be one candidate. By splitting the change into smaller PRs, that will make it easier to review piece by piece, and make the overall experience smoother both for you as a new contributor receiving feedback from reviewers, as well as making the reviewers' jobs easier.

MrFreezeex commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -514,6 +514,12 @@ data:
+  # Default global namespace behavior for ClusterMesh namespace filtering.
+  # When namespace-based filtering is active, determines default behavior for non-annotated namespaces.
+  clustermesh-default-global-namespace: "{{ .Values.clustermesh.defaultGlobalNamespace }}"
+{{- end }}
```
The values "clustermesh.defaultGlobalNamespace" is always defined so hasKey will always be true and can be removed here

MrFreezeex commented on NAMESPACE_EXPORT_CONTROL_TESTING.md:
```diff
@@ -0,0 +1,229 @@
+# Namespace-based Export Control Testing Infrastructure
```
We typically don't really document test procedure about new features like that, could you remove this?

krunaljain committed:
feat(clustermesh): Implement namespace watcher infrastructure for global service filtering

This commit introduces the foundational namespace watcher infrastructure required
for CFP-39876 global service dual requirements implementation.

Major elements included:
- pkg/clustermesh/namespace_watcher.go - Core namespace tracking logic
- pkg/clustermesh/namespace_watcher_cell.go - Hive cell integration
- pkg/clustermesh/cell.go - Cell registration
- pkg/annotation/k8s.go - GlobalNamespace annotation constant
- pkg/annotation/clustermesh.go - Namespace filtering helper

Includes original commits:
- Core namespace watcher functionality from multiple original commits
- Cell architecture setup and configuration management

Signed-off-by: krunaljain <krunal.bitsy@gmail.com>

krunaljain committed:
feat(clustermesh): Implement CiliumEndpoints, CiliumIdentities, CiliumEndpointSlice synchronization with namespace filtering

This commit adds namespace-aware synchronization for Cilium resources in clustermesh-apiserver,
building on the namespace watcher infrastructure from commit 1.

Major elements included:
- clustermesh-apiserver/clustermesh/converters.go - Namespace filtering for resource converters
- clustermesh-apiserver/clustermesh/k8s/resource_ctors.go - Resource constructors with indexers
- clustermesh-apiserver/clustermesh/k8s/resources.go - Namespace resource integration
- clustermesh-apiserver/clustermesh/synchronizer.go - Namespace change processors
- clustermesh-apiserver/clustermesh/cells.go - Namespace watcher cell integration

Includes original commits:
- Resource converter enhancements for namespace filtering
- Indexer setup for efficient namespace-based lookups
- Processor registration for dynamic namespace changes

Signed-off-by: krunaljain <krunal.bitsy@gmail.com>

krunaljain committed:
feat(clustermesh): Implement global service dual requirements in service sync and endpoint slice sync

This commit implements the core global service logic with dual requirements (service annotation + namespace filtering)
for non-MCS scenarios, building on the previous namespace watcher and resource synchronization infrastructure.

Major elements included:
- operator/watchers/service_sync.go - Enhanced service converter with namespace filtering
- pkg/annotation/clustermesh.go - Complete dual requirement validation helpers
- pkg/clustermesh/endpointslicesync/*.go - Namespace-aware endpoint slice synchronization
- Service synchronization respects both global annotation and namespace status

Includes original commits:
- Service sync converter enhancements for CFP-39876 dual requirements
- Endpoint slice sync integration with namespace filtering
- Annotation helpers for dual requirement validation

Signed-off-by: krunaljain <krunal.bitsy@gmail.com>

krunaljain committed:
feat(clustermesh): Implement MCS API dual requirements validation

This commit implements Multi-Cluster Services (MCS) API support with dual requirements validation,
ensuring ServiceExport/ServiceImport operations respect both service annotation and namespace filtering.

Major elements included:
- pkg/clustermesh/mcsapi/cell.go - Namespace tracker integration in MCS API cell
- pkg/clustermesh/mcsapi/serviceimport_controller.go - Enhanced controller with namespace validation
- ServiceExport validation rejects services without global annotation
- ServiceImport creation respects global namespace requirements
- Dual requirement enforcement for MCS API operations

Includes original commits:
- MCS API controller enhancements for CFP-39876 compliance
- ServiceExport/ServiceImport validation with namespace checking
- Integration with clustermesh namespace filtering infrastructure

Signed-off-by: krunaljain <krunal.bitsy@gmail.com>

krunaljain committed:
feat(clustermesh): Add comprehensive E2E tests and configuration for global service dual requirements

This commit completes the global service dual requirements implementation by adding comprehensive
end-to-end testing, configuration templates, and validation tests for the full CFP-39876 feature set.

Major elements included:
- Test files: *_test.go, testdata/* - Comprehensive unit and integration tests
- Workflow files: .github/workflows/conformance-clustermesh.yaml - CI/CD integration
- Configuration: install/kubernetes/cilium/templates/* - Helm templates and values
- Test fixtures: contrib/testing/kind-* - Local testing configuration
- Complete implementations: service_informer.go, serviceimport_controller.go with full dual requirement logic

Testing coverage:
- Namespace watcher functionality and edge cases
- Service sync with dual requirements validation
- MCS API controller dual requirement enforcement
- End-to-end clustermesh scenarios with namespace filtering
- Backward compatibility and configuration edge cases

Includes original commits:
- All testing infrastructure and validation
- Configuration templates and deployment updates
- Complete feature implementation validation
- CI/CD workflow integration for testing dual requirements

Signed-off-by: krunaljain <krunal.bitsy@gmail.com>
